### PR TITLE
Add support for stop conditions

### PR DESCRIPTION
**Pull request by danieldk**
## Description

This change introduces a new base class, StopCondition. Implementations of this class provide stop conditions for generation. A stop condition gets the most recent parse state and updates one of two arrays that mark generation of a sequence as completed:

- completed_exclude: for sequences that are completed.
- completed_include: for sequences that are completed and for which the last generated piece should still be emitted.

Initially, there are two stop conditions:

- EndOfSequenceCondition: stop generating a sequence when the end-of-sequence piece id is predicted. This replaces the previous manual handling of the end-of-sequence piece id.
- MaxNewPiecesCondition: stop generating when the maximum number of generation steps has been reached.

The generator and generator state has been adjusted to work with stop conditions.

### Types of change

Feature

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: curated_transformers/generation/state.py [hunk 1]
```diff
@@ -66,11 +110,13 @@ def _remove_completed(self, completed: Tensor):
...
         """
         not_completed = completed.logical_not()
+        self.generated_ids = self.generated_ids[not_completed]
         self.attention_mask = self.attention_mask[not_completed]
         if self.cache is not None:
             self.cache = [
                 layer_cache.filter_batch_items(not_completed)
                 for layer_cache in self.cache
             ]
+        self.prompt_ids = self.prompt_ids[not_completed]
```
Comment (shadeMe) [inline curated_transformers/generation/state.py:120]: We don't really need to update `prompt_ids`, right? This is just to keep the shapes consistent, I take?

File: curated_transformers/generation/stop.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
+from abc import ABC, abstractmethod
```
Comment (shadeMe) [inline curated_transformers/generation/stop.py:1]: Why not call the module `stop_conditions.py`?

File: curated_transformers/generation/stop.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    @abstractmethod
+    def update_completed(
+        self,
+        *,
+        state: "GeneratorState",
+        completed_exclude: Tensor,
+        completed_include: Tensor
+    ):
+        """
+        Update completed sequences according to the stop condition.
```
Comment (shadeMe) [inline curated_transformers/generation/stop.py:22]: Let's add a note to indicate the `Tensor` arguments are `out` parameters/modified in-place (not immediately clear currently).

File: curated_transformers/generation/dolly_v2.py [hunk 4]
```diff
@@ -52,6 +53,13 @@ def from_hf_hub(
         return cls(tokenizer, causal_lm)
 
     def generate(self, prompts: List[str], config: GeneratorConfig) -> List[str]:
+        # Fill config when necessary.
+        eos_id = self.eos_id if config.eos_id is None else config.eos_id
+        max_new_pieces = 256 if config.max_new_pieces is None else config.max_new_pieces
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/generation/dolly_v2.py]: It's not arbitrary, it's what the upstream Dolly v2 generation model uses.

File: curated_transformers/generation/state.py [hunk 5]
```diff
@@ -13,40 +14,64 @@ class GeneratorState(Generic[CacheT]):
...
     seq_ids: Tensor
+    prompt_ids: Tensor
+    generated_ids: Tensor
 
     def __init__(
-        self,
-        *,
-        attention_mask: Tensor,
-        cache: Optional[List[CacheT]],
+        self, *, attention_mask: Tensor, cache: Optional[List[CacheT]], ids: Tensor
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/generation/state.py]: Fixed.

File: curated_transformers/generation/state.py [hunk 6]
```diff
@@ -13,40 +14,64 @@ class GeneratorState(Generic[CacheT]):
...
     ) -> None:
         self.attention_mask = attention_mask
         self.positions = attention_mask.int().cumsum(-1) - 1
         self.cache = cache
         self.seq_ids = torch.arange(
             0, self.attention_mask.size(0), device=attention_mask.device
         )
+        self.prompt_ids = ids
+        self.generated_ids = torch.zeros(
+            (ids.size(0), 0), dtype=ids.dtype, device=ids.device
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/generation/state.py]: Fixed.

File: curated_transformers/generation/state.py [hunk 7]
```diff
@@ -13,40 +14,64 @@ class GeneratorState(Generic[CacheT]):
...
+        )
 
     @property
     def is_finished(self):
         return len(self.seq_ids) == 0
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/generation/state.py:60]: Added.

File: curated_transformers/generation/stop.py (hunk 2 above)
Comment (danieldk) [reply curated_transformers/generation/stop.py]: Good idea, fixed.

File: curated_transformers/generation/stop.py (hunk 3 above)
Comment (danieldk) [reply curated_transformers/generation/stop.py]: Updated docstring.

File: curated_transformers/generation/config.py [hunk 8]
```diff
@@ -22,6 +46,20 @@ def logits_transform(self) -> LogitsTransform:
         """
         ...
 
+    def stop_condition(self) -> StopCondition:
+        conditions = CompoundStopCondition()
```
[the remark this answers is not in the document]
Comment (shadeMe) [reply curated_transformers/generation/config.py:49]: Makes sense :+1:

**Review by shadeMe — APPROVED** [review]
LGTM! :tada:

**Pull request merged**